cpu/stm32/dist/irqs/gen_irqs.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2020 Inria
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+"""Generate a header with the number of IRQs for a given STM32 family."""
+
+import os
+import argparse
+import re
+
+
+CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
+RIOTBASE = os.getenv(
+    "RIOTBASE", os.path.abspath(os.path.join(CURRENT_DIR, "../../../..")))
+STM32_INCLUDE_DIR = os.path.join(RIOTBASE, "cpu/stm32/include")
+STM32_CMSIS_INCLUDE_DIR = os.path.join(
+    RIOTBASE, STM32_INCLUDE_DIR, "vendor/cmsis/{}/Include/")
+STM32_CMSIS_HEADER = os.path.join(
+    RIOTBASE, STM32_CMSIS_INCLUDE_DIR, "{}.h")
+STM32_IRQS_DIR = os.path.join(
+    RIOTBASE, STM32_INCLUDE_DIR, "irqs/{}/irqs.h")
+
+IRQS_FORMAT = """
+/*
+ * PLEASE DON'T EDIT
+ *
+ * This file was automatically generated by
+ * ./cpu/stm32/dist/irqs/gen_irqs.py
+ */
+
+#ifndef IRQS_{cpu_fam}_H
+#define IRQS_{cpu_fam}_H
+
+#ifdef __cplusplus
+extern "C" {{
+#endif
+
+{irq_lines}
+
+#ifdef __cplusplus
+}}
+#endif
+
+#endif /* IRQS_{cpu_fam}_IRQ_H */
+"""
+
+
+def list_cpu_lines(cpu_fam):
+    """Returns the list CPU lines for a given family"""
+    headers = os.listdir(STM32_CMSIS_INCLUDE_DIR.format(cpu_fam))
+    headers.remove("stm32{}xx.h".format(cpu_fam))
+    headers.remove("system_stm32{}xx.h".format(cpu_fam))
+    return [header.split(".")[0] for header in headers]
+
+
+def irq_numof(cpu_fam, cpu_line):
+    """Parse the CMSIS to get the list IRQs."""
+    cpu_line_cmsis = STM32_CMSIS_HEADER.format(cpu_fam, cpu_line)
+
+    with open(cpu_line_cmsis, 'rb') as cmsis:
+        cmsis_content = cmsis.readlines()
+
+    for line_idx, line in enumerate(cmsis_content):
+        try:
+            line = line.decode()
+        except UnicodeDecodeError:
+            # skip line that contains non unicode characters
+            continue
+        # skip useless lines
+        if (
+            # Skip lines with comments
+            line.startswith("/*") or
+            # Skip line starting with too many spaces
+            line.startswith(" " * 8) or
+            # Skip remapped USB interrupt
+            # (set as alias for USBWakeUp_IRQn on stm32f3)
+            "USBWakeUp_RMP_IRQn" in line
+        ):
+            continue
+        # Stop at the end of the IRQn_Type enum definition
+        if "IRQn_Type" in line:
+            break
```

Comment:
Still didn't answer why not to do this the other war around.

Reply:
see https://github.com/RIOT-OS/RIOT/pull/14595#discussion_r481213105